design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,217 @@
+Ironic community
+([hw config](https://review.opendev.org/c/openstack/ironic-specs/+/740721))
+to export and import an entire system iDRAC inventory
+including RAID and networking requirements. Compared to that approach, this
+proposal attempts a more lightweight, vendor-neutral approach just for BIOS
+configuration.
+
+### Goals
+
+Provide a vendor-neutral approach of updating BIOS configuration in a bulk
+manner across all similar hardware. Rely on the user to create the bulk input
+from cached BIOS settings on a machine. Give access to the BIOS registry which
+can be used as documentation and reference in creating the input.
+
+### Non-Goals
+
+This proposal does not attempt to fulfill the following goals:
+
+- Allow the transfer of BIOS configuration between different vendor’s
+  servers or different vendor models
+- Support bulk set of other configurations besides BIOS
+- Provide storage mechanism for BIOS settings
+- Filter or validate BIOS settings
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+When provisioning a large amount of machines with identical settings, our
+customers want to provide a known, validated payload with all the desired
+firmware configuration.
+
+This ensures all the machines will have the identical settings, reducing issues
+caused by unexpected configuration differences.
+
+Since the name/value pair mapping is vendor specific and not intuitive, it
+doesn’t lend itself to creating the input by hand using existing vendor
+documentation, except for potentially a small subset of names. It’s desirable
+to take an existing BIOS config set and modify the current settings to the
+desired settings. The BIOS Registry can be used as a source of documentation
+to modify the settings.
+
+## Design Details
+
+This proposes three new data elements in the BareMetalHost - the current BIOS
+settings retrieved from Ironic, the BIOSRegistry read from Ironic (effectively
+a schema for the BIOS settings), and a bulk BIOS settings for Ironic to write
+to the BMC. The current BIOS settings and BIOS Registry can be used to create
+the bulk BIOS settings which can be set on multiple hosts of the same type,
+the current settings and registry will not be used on these hosts.
+
+The current BIOS settings are retrieved from the BMC by Ironic and cached
+whenever the node moves to `manageable` or `cleaning`. These BIOS settings
+can be read via the Ironic API (`/v1/nodes/{node_ident}/bios`) as a set of
+name:value pairs. Ironic can set the BIOS Configuration values through the use
+of clean_steps as part of the manual cleaning process as described in the Ironic
+documentation. BIOS settings are configured as name/value pairs and can include
+from one to multiple pairs.
+
+What we are proposing is to make the cached BIOS settings from Ironic available
+to the user. This can be either via a ConfigMap or special purpose CRD
+referenced in the BareMetalHostStatus.  As there are no fields in BIOS
```

Comment:
I think what we'll want is a new CRD with a name like `BIOSRegistry` that has a host reference in the spec fields. When one is created, a controller would wait for ironic to have the information it needs and then fill in information in the status fields of `BIOSRegistry`. That means the main reconcile loop of the baremetal-operator doesn't have to deal with this new feature at all, and there can be separate RBAC settings for being able to ask for BIOS info.

We could (eventually) do things like filter out read-only values, or even let the user ask for specific settings by name.

What do you think, @zaneb?

Reply:
I hadn't really thought about a separate CRD, but I like the idea. We ended up kinda regretting not doing that with HardwareDetails, so it might make sense to get ahead of that.
The relationship is 1:1 so I think it should be linked just by having the same name so that the lookup is O(1) in both directions, rather than having to list and filter all of them to find the one that corresponds to a particular host. That's one thing a CRD makes possible that you can't really do with a ConfigMap, since the user can create their own ConfigMaps with arbitrary names for other reasons.
This could be done by a separate controller, but I'm not sure we want two things talking to ironic. A new CRD could just as easily be created by the existing BMH controller, and this would ensure it got updated at the right points in the workflow.
For the Registry, having it in a CRD also allows us to have a(n OpenAPI) schema for the schema.

On the flip side:
This paragraph is talking about exposing the actual settings, but the name `BIOSRegistry` implies that you're talking about the schema. Are you suggesting to combine them into a single CRD?
We know that the registry (i.e. schema) is very large - on the same order of magnitude as the limit on k8s object size - and we talked about compressing it before storing in a ConfigMap. That wouldn't really be an option for a CRD.

Reply:
I was suggesting 2 CRDs. One to ask for the current settings, and another to provide a set of values to change for a particular host. The current settings CRD could have a subset of the info from the registry. The change CRD would only need a list of name/value pairs.

Reply:
The current settings will come directly from Ironic via the /v1/nodes/{node_ident}/bios API and will only include name/value pairs (https://docs.openstack.org/ironic/rocky/admin/bios.html) including RO attributes.  We could modify this based on the contents of BIOSRegistry (including filtering to remove RO atrributes) but I thought otherwise we'd keep it as is, copy it to the change CRD, and update the values there based on the allowed values in the BIOSRegistry.  This change CRD would provide the set for values to use for hosts.